interpreter.go

```diff
@@ -156,35 +272,216 @@ func makeCallStack(limit int) callStack {
 // TODO(dcunnin): Add multi output.
 
 type interpreter struct {
-	Stack        callStack
-	ExternalVars vmExtMap
+	stack          callStack
+	idArrayElement identifier
+	idInvariant    identifier
+	externalVars   vmExtMap
+}
+
+func (i *interpreter) capture(freeVars identifiers) bindingFrame {
+	var env bindingFrame
+	for _, fv := range freeVars {
+		env[fv] = i.stack.lookUpVar(fv)
+	}
+	return env
+}
+
+type fieldHideMap map[string]astObjectFieldHide
+
+func (i *interpreter) objectFieldsAux(obj value) fieldHideMap {
+	r := make(fieldHideMap)
+	switch obj := obj.(type) {
+	// case *valueExtendedObject:
+	// TODO(dcunnin): this
+
+	case *valueSimpleObject:
+		for fieldName, field := range obj.fields {
+			r[fieldName] = field.hide
+		}
+
+		// case *valueComprehensionObject:
+		// TODO(dcunnin): this
+	}
+	return r
+}
+
+func (i *interpreter) objectFields(obj value, manifesting bool) []string {
+	var r []string
+	for fieldName, hide := range i.objectFieldsAux(obj) {
+		if !manifesting || hide != astObjectFieldHidden {
+			r = append(r, fieldName)
+		}
+	}
+	return r
 }
 
-func (i *interpreter) execute(a astNode) (value, error) {
+func (i *interpreter) findObject(f string, curr value, startFrom int, counter *int) value {
+	switch curr := curr.(type) {
+	// case *valueExtendedObject:
+	// TODO(dcunnin): this
+
+	case *valueSimpleObject:
+		if *counter >= startFrom {
+			if _, ok := curr.fields[f]; ok {
+				return curr
+			}
+		}
+		*counter++
+
+		// case *valueComprehensionObject:
+		/*
+			if *counter >= startFrom {
+				// TODO(dcunnin): this
+			}
+			*counter++
+		*/
+	}
+	return nil
+}
+
+func (i *interpreter) objectIndex(loc *LocationRange, obj value, f string, offset int) (astNode, error) {
+	var foundAt int
+	self := obj
+	found := i.findObject(f, obj, offset, &foundAt)
+	if found == nil {
+		return nil, makeRuntimeError(loc, fmt.Sprintf("Field does not exist: %s", f))
+	}
+	switch found := found.(type) {
+	case *valueSimpleObject:
+		field := found.fields[f]
+		i.stack.newCall(loc, found, self, foundAt, found.upValues)
+		return field.body, nil
+	// case *valueComprehensionObject:
+	/*
+		// TODO(dcunnin): this
+	*/
+	default:
+		return nil, fmt.Errorf("Internal error: findObject returned unrecognized type: %s", reflect.TypeOf(found))
+	}
+}
+
+func (i *interpreter) evaluate(a astNode) (value, error) {
 	// TODO(dcunnin): All the other cases...
 	switch ast := a.(type) {
+	case *astArray:
+		self, offset := i.stack.getSelfBinding()
+		var elements []*thunk
+		for _, el := range ast.elements {
+			elThunk := makeThunk(i.idArrayElement, self, offset, el)
+			elThunk.upValues = i.capture(el.FreeVariables())
+			elements = append(elements, elThunk)
+		}
+		return &valueArray{elements}, nil
+
 	case *astBinary:
 		// TODO(dcunnin): Assume it's + on numbers for now
-		leftVal, err := i.execute(ast.left)
+		leftVal, err := i.evaluate(ast.left)
 		if err != nil {
 			return nil, err
 		}
 		leftNum := leftVal.(*valueNumber).value
```

Comment:
This type assertion will panic if this thing isn't a `valueNumber`.  Might want to check it and throw an error instead.